### ISSUE TITLE
VSCode Breakpoints not working

### ISSUE THREAD

**Issue by adim1973** [OfficeDev/Excel-Custom-Functions#336]
I used the Yeoman generator for generating an Excel Custom Functions Add-in project, with script type: TypeScript and I'm trying to debug it using VSCode. Any breakpoints set appear as unbound, plus there's a "read source map" error (which might be related?).

<img width="768" alt="error" src="https://user-images.githubusercontent.com/5249431/178109632-9a1dd079-10c7-4da8-abf6-67c6242a3d95.png">

- I followed the steps in the [Custom functions debugging tutorial ](https://docs.microsoft.com/en-US/office/dev/add-ins/excel/custom-functions-debugging#use-the-vs-code-debugger-for-excel-desktop) and didn't change anything in the generated code.
- I chose TypeScript because of this recommendation in the debugging tutorial.
<img width="475" alt="image" src="https://user-images.githubusercontent.com/5249431/178109887-ddf72260-a0c9-412e-9ac1-f47a1f883702.png">

I found this related closed issue (https://github.com/OfficeDev/Excel-Custom-Functions/issues/236), but everything seems OK with the foundation.win32.bundle.

Any ideas?

Adrian

**Comment by LuyiH** [comment]
@yaweizhu-henson  please help with this issue, thanks

**Comment by igor-ribeiiro** [comment]
Hello @adim1973, the tutorial might be a little wrong as I also got some problems in getting the breakpoints at first.
Could you try running the dev-server first, wait for it to complete, and then click on the debugging option `Excel Desktop (Custom Functions)`?

If it works, please let me know and I will work to also update the documentation to reflect that.

**Comment by akrantz** [comment]
When you use "npm start" or VS Code start debugging, it will check if the dev-server is running. If not running, it will start it, or give an error if it cannot start it. It waits for it to be ready, before doing the sideload.

If there's an issue with breakoints in VS Code, if using the Edge webview, it's more likely that it's not binding the breakoints at start up. You can put a debugger statement in your code to ensure that it breaks and then see if it can bind the breakpoints after that. You can also use the VS Code command to diagnose breakpoint issues. (Let me find the link to VS Code docs and I'll update this comment.)

**Comment by adim1973** [comment]
Still not working.

<img width="963" alt="image" src="https://user-images.githubusercontent.com/5249431/180958343-d5a7b9df-ba71-4809-81b6-9e523628b271.png">

**Comment by adim1973** [comment]
I replaced
```
"resolveSourceMapLocations": [
        "${workspaceFolder}/**",
        "!**/node_modules/**"
    ]
```
with
`'"resolveSourceMapLocations": null'`
but no change.

<img width="960" alt="image" src="https://user-images.githubusercontent.com/5249431/182320483-9cf6d26b-1f53-45b7-9a8e-eeee860aab7f.png">

**Comment by millerds** [comment]
Are you still getting the same errors about resource maps in the debug console?

**Comment by adim1973** [comment]
No, the "read source map" error is gone.

**Comment by igor-ribeiiro** [comment]
@adim1973 I investigated the bug and it is only happening in JavaScript Custom Functions right now. It should be working for TypeScript. Could you try using typescript in the meanwhile? I will be working on fixing this issue for the JavaScript template.

**Comment by adim1973** [comment]
@igor-ribeiiro I'm using the TypeScript template, as mentioned in the first post.

**Comment by igor-ribeiiro** [comment]
Then the debugger should be working (just tested now). Could you testing the following:
Could you please try the following:
1.  yo office
2. Select Excel Custom Functions then Typescript as they come up
3. Put some breakpoints on  src/functions.ts inside the `clock` function for example
4. On a terminal, run `npm run dev-server` and wait for it to finish building (don't close it, leave it running even after it is built).
5. Open the project on VSCode. Select the Excel Desktop (Custom Functions) debugger option and run it
6. On Excel, select a cell and write: `=contoso.clock()`
7. The breakpoint should be hit on VSCode

**Comment by adim1973** [comment]
[test.7z.zip](https://github.com/OfficeDev/Excel-Custom-Functions/files/9308323/test.7z.zip)

Done. Same issue. See attached a video with the whole steps (it's an mp4 archived in 7z format, with an extra .zip extension, for compliance).

**Comment by igor-ribeiiro** [comment]
Thanks for the patience @adim1973, I saw the video and I think you are running on a known issue that sometimes happens (we can't consistently reproduce on all machines yet) in which Excel CFs won't properly load if you didn't wait for the dev-server to completely built.

Other than that, please try clearing the WEF cache as it can bug when multiple add-ins have the same function names (such as when creating multiple templates). That is also something we will follow-up on fixing.
Please, delete the following folder: `C:\Users\<username>\AppData\Local\Microsoft\Office\16.0\Wef`

Before clicking on the debugger `Excel Desktop (Custom Functions)`, could you wait for the dev-server to finish serving completely? You should wait until a message saying it compiled successfully appears:
![image](https://user-images.githubusercontent.com/18221895/184246630-3b6d5e45-af32-4d7e-85b1-a1867062d262.png)

Let me know if this solved the issue, we will investigate why these bugs are happening.

**Comment by adim1973** [comment]
@igor-ribeiiro Cleared wef folder, waited until dev-server finished, still same issue with the breakpoint. Plus the following in debug console (see video):
```
Could not read source map for foundation.win32.bundle: Invalid URL
Could not read source map for file:///C:/Program%20Files/Microsoft%20Office/Root/Office16/excel-udf-host.win32.bundle: ENOENT: no such file or directory, open 'c:\Program Files\Microsoft Office\Root\Office16\excel-udf-host.win32.bundle.map'
``
[test.7z.zip](https://github.com/OfficeDev/Excel-Custom-Functions/files/9315831/test.7z.zip)
`

**Comment by adim1973** [comment]
@akrantz already did that, as you can see in the video.

**Comment by igor-ribeiiro** [comment]
> @akrantz already did that, as you can see in the video.

Now the main thing that could be happening is you could be using an old version of Office. Can you please try updating your Office to the latest?

**Comment by akrantz** [comment]
It would be helpful to understand what the url is for the breakpoint(s) and why that doesn't resolve to a loaded script file. VS Code has support for a command to troubleshoot this info when debugging. 

You can confirm that the .map file is copied into the wef cache location, but that should be happening. The functions.js bundle and function.js.map file should be in the same directory.

If this is available, then the any issues with binding the breakpoints have to do with the debugger (VS Code). 

You might also first see if you can set a breakpoint in the javascript bundle directly, or put a debugger statement in code to force breaking in that bundle. The last line of that js file should have a comment for the source map, which should point to the map file in the same directory. That map file will contain a list of source files near the beginning, and VS Code should be able to map a loaded script file to one of those paths.

**Comment by akrantz** [comment]
This might help: https://code.visualstudio.com/docs/nodejs/nodejs-debugging#_javascript-source-map-tips

The command "Debug: Diagnose Breakpoint Problems" can help show info about why a breakpoint request may not be binding to a loaded script.

**Comment by adim1973** [comment]
OK, I'll follow those hints and see if it the issue can be solved.
You can close this, I will open another issue with new info, if needed.

**Comment by igor-ribeiiro** [comment]
Hey @adim1973, we found a bug in our debugger. The following PR is addressing that: https://github.com/OfficeDev/Excel-Custom-Functions/pull/340

While it is not merged yet, feel free to try it beforehand manually. It is just a one-line change that should help you debug.
Please let me know if the PR fixed the issue for you. Thanks for your patience, you helped us identify a real problem in our tooling!

**Comment by adim1973** [comment]
@igor-ribeiiro I can confirm that deleting the line `devtoolModuleFilenameTemplate: "webpack:///[resource-path]?[loaders]", `in file webpack.config.js fixes the VS code breakpoints not working issue. Well done!
Did I understand correctly from the comments of the PR [#340](https://github.com/OfficeDev/Excel-Custom-Functions/pull/340) that debugging directly in VS code for JavaScript Office add-ins (not TypeScript) will also be possible?

**Comment by igor-ribeiiro** [comment]
@adim1973 Yes, it would also be possible in the near future! We are working on publishing the fix!
Thanks for the patience, I will be closing the issue!
